### ISSUE TITLE
Massive performance issues for all Pylance features in some projects

### ISSUE THREAD

**Issue by MiguelGuthridge** [microsoft/pylance-release#3632]
## Environment data

-   Language Server version: v2022.11.20
-   OS and version: Fedora 36
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.8
-   VS Code version: 1.73.1

## Code Snippet

This issue only seems to happen for particularly large projects, so I can't give a small code sample.

This is a reasonably small repo that I have experienced issues with.

https://github.com/MiguelGuthridge/Jestspectation

I have experienced similar issues with another project that has this as a dependency. As it is a university assignment, I sadly can't share the code for it yet, however once it is due in a few days I should be able to share a link.

If there are any specific things I can do  to help narrow down the issue, I would be happy to help time permitting.

## Repro Steps

1. Work on writing code normally. There doesn't appear to be any specific action that causes the issue.
2. Keep working for roughly 10-15 minutes.

## Expected behavior

The following features perform well:
* Predictions
* Documentation popups
* Go to definition
* Typing (in particular hitting enter is slower than other keys, which leads to things getting typed in the wrong place)
* Inline errors from linters

## Actual behavior

The features listed above are extremely slow, often taking from seconds to minutes to complete.

## Logs
It's long.

[pylance.txt](https://github.com/microsoft/pylance-release/files/10013918/pylance.txt)

I would embed it here but it lagged my browser a ton owing to it being >2k lines long for just over 1 minute of usage.

**Comment by erictraut** [comment]
Based on the logs, it looks like a file system change event in the ".pytest_cache" directory is triggering a re-indexing of all files, including everything in the site-packages directory. Because this is only a partial log, I can't say for sure whether this is the cause of the apparent slow-down, but it could explain the symptoms.

**Comment by MiguelGuthridge** [comment]
Is there anything I can do to help investigate this further? The only workaround I've found is to spend 30 seconds reloading the window every 3-5 minutes, which has quite a huge impact on my workflow. Based on what you've said, one commonality between both of the projects that I am experiencing these issues with is their use of the `jestspectation.pytest` module as a Pytest plugin - perhaps the system for discovering Pytests is causing issues when this plugin gets loaded?

**Comment by rchiodo** [comment]
A set of distinct repo steps would help. I'm trying with your `jestspectation` repository and I can't repro a problem.

Is there something in particular you are doing in that repository? 

I tried running tests, doing completions, hovers, goto defs, etc.

**Comment by MiguelGuthridge** [comment]
I can't think of anything in particular.

I just reproduced it in my assignment code by moving a function from one file to another and adding some aliases to the imports. I've also triggered it by writing new tests, hitting run (with debugging). In general the only commonality seems to be time - after keeping it open for a certain amount of time it gets quite slow. Owing to the ridiculous requirements of the assignment I was given I haven't had the time to check how it goes for other projects - I'll hopefully be able to get back to you on that over the weekend.

**Comment by MiguelGuthridge** [comment]
I'm starting to think this might not be a Pylance issue. Although I hadn't experienced any issues with other languages until today, I recently experienced the same symptoms when working with Rust. To double check, I tried with all my Python-related extensions disabled, but it was still happening. It could be entirely unrelated issues, but the outcomes were similar enough that I've opened an issue on the main VSCode repo

**Comment by heejaechang** [comment]
There is one issue we can improve which we noticed from your log. that will be included in our next release. 

What we found from your log is this

```
[Info  - 1:55:25 am] (967618) SourceFile: Received fs event 'add' for path '/home/migue/Source/Ensemble/ensemble_request/__pycache__/helpers.cpython-310.pyc'
[Info  - 1:55:25 am] (967618) SourceFile: Received fs event 'add' for path '/home/migue/Source/Ensemble/.pytest_cache/v'
(967618) Background analysis message: invalidateAndForceReanalysis
```

basically, external components are touching some files/folders under your workspace root, and we are responding to those. We added some more rules/cases to ignores when something unknown happens under the root.

**Comment by MiguelGuthridge** [comment]
That's nice to hear! I think I figured out what the major issue is though - the Todo Tree extension seems to be triggering Rust Analyzer and Pylance to constantly rescan the entire project directory. It's great to see that this will still result in improvements to Pylance too though!

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2022.11.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20221141-30-November-2022-PreRelease)
